File: CFSModeRuleManager.js
var RULETYPE_NONE = -1;
var RULETYPE_NORMAL = 0;
var RULETYPE_LEFT_FIVE = 1;
var RULETYPE_DIVIDE = 2;
var RULETYPE_DOUBLE = 3;
var RULETYPE_LEFT_SAME = 4;
var RULETYPE_FIVE = 5;
var RULETYPE_TTS = 6;
var RULETYPE_CONTINUE = 7;

var CFSModeRuleManager = function () {};

CFSModeRuleManager.getPaiGowRuleStyle = function (paiGowArray) {
    if (paiGowArray.length != 3) {
        return RULETYPE_NONE;
    }
    var paiGowNumArray = new Array();
    for (var i = 0; i < paiGowArray.length; i++) {
        var p = paiGowArray[i];
        paiGowNumArray.push(p.pg_preNum);
        paiGowNumArray.push(p.pg_postNum);
    }
    if (CFSModeRuleManager.isRuleTypeContinue(paiGowNumArray)) {
        return RULETYPE_CONTINUE;
    }
    if (CFSModeRuleManager.isRuleTypeTTS(paiGowNumArray)) {
        return RULETYPE_TTS;
    }
    if (CFSModeRuleManager.isRuleTypeFive(paiGowNumArray)) {
        return RULETYPE_FIVE;
    }
    if (CFSModeRuleManager.isRuleTypeDouble(paiGowNumArray)) {
        return RULETYPE_DOUBLE;
    }
    if (CFSModeRuleManager.isRuleTypeDivide(paiGowNumArray)) {
        return RULETYPE_DIVIDE;
    }
    if (CFSModeRuleManager.isRuleTypeLeftSame(paiGowNumArray)) {
        return RULETYPE_LEFT_SAME;
    }
    if (CFSModeRuleManager.isRuleTypeLeftFive(paiGowNumArray)) {
        return RULETYPE_LEFT_FIVE;
    }
    if (CFSModeRuleManager.isRuleTypeNORMAL(paiGowNumArray)) {
        return RULETYPE_NORMAL;
    }
    return RULETYPE_NONE;

};

CFSModeRuleManager.getPaiGowCountInArray = function (paiGowArray, cardValue) {
    var count = 0;
    for (var i = 0; i < paiGowArray.length; i++) {
        if (paiGowArray[i] == cardValue) count++;
    }
    return count;
};

CFSModeRuleManager.cardValueInArray = function (paiGowArray, cardCount) {
    for (var i = 0; i < paiGowArray.length; i++) {
        var count = CFSModeRuleManager.getPaiGowCountInArray(paiGowArray, paiGowArray[i]);
        if (count == cardCount) return paiGowArray[i];
    }
    return 0;
};

CFSModeRuleManager.isRuleTypeTTS = function (paiGowNumArray) {
    if (CFSModeRuleManager.getPaiGowCountInArray(paiGowNumArray, 2) != 2) return false;
    if (CFSModeRuleManager.getPaiGowCountInArray(paiGowNumArray, 3) != 2) return false;
    if (CFSModeRuleManager.getPaiGowCountInArray(paiGowNumArray, 6) != 2) return false;
    return true;
};

CFSModeRuleManager.isRuleTypeDouble = function (paiGowNumArray) {
    if (CFSModeRuleManager.getPaiGowCountInArray(paiGowNumArray, 1) != 2) {
        if (CFSModeRuleManager.getPaiGowCountInArray(paiGowNumArray, 4) != 2) return false;
        if (CFSModeRuleManager.getPaiGowCountInArray(paiGowNumArray, 5) != 2) return false;
        if (CFSModeRuleManager.getPaiGowCountInArray(paiGowNumArray, 6) != 2) return false;
        return true;
    }
    if (CFSModeRuleManager.getPaiGowCountInArray(paiGowNumArray, 2) != 2) {
        if (CFSModeRuleManager.getPaiGowCountInArray(paiGowNumArray, 4) != 2) return false;
        if (CFSModeRuleManager.getPaiGowCountInArray(paiGowNumArray, 5) != 2) return false;
        if (CFSModeRuleManager.getPaiGowCountInArray(paiGowNumArray, 6) != 2) return false;
        return true;
    }
    if (CFSModeRuleManager.getPaiGowCountInArray(paiGowNumArray, 3) != 2) {
        if (CFSModeRuleManager.getPaiGowCountInArray(paiGowNumArray, 4) != 2) return false;
        if (CFSModeRuleManager.getPaiGowCountInArray(paiGowNumArray, 5) != 2) return false;
        if (CFSModeRuleManager.getPaiGowCountInArray(paiGowNumArray, 6) != 2) return false;
        return true;
    }
    return true;
};

CFSModeRuleManager.isRuleTypeFive = function (paiGowNumArray) {
    for (var i = 0; i < paiGowNumArray.length; i++) {
        if (CFSModeRuleManager.getPaiGowCountInArray(paiGowNumArray, paiGowNumArray[i]) == 5)  return true;
    }
    return false;
};

CFSModeRuleManager.isRuleTypeContinue = function (paiGowNumArray) {
    for (var i = 1; i < 7; i++) {
        if (!paiGowNumArray.contains(i)) return false;
    }
    return true;
};

CFSModeRuleManager.isRuleTypeNORMAL = function (paiGowNumArray) {
    var cardValue = CFSModeRuleManager.cardValueInArray(paiGowNumArray, 3);
    if (cardValue == 0) return false;
    var count = 0;
    for (var i = 0; i < paiGowNumArray.length; i++) {
        if (paiGowNumArray[i] != cardValue) count += paiGowNumArray[i];
    }
    var criterionValue = (GameDataManager.gameLevel == GAMELEVEL_EASY) ? 13 : 14;
    return (count >= criterionValue) ? true : false;
};

CFSModeRuleManager.isRuleTypeLeftFive = function (paiGowNumArray) {
    var cardValue = CFSModeRuleManager.cardValueInArray(paiGowNumArray, 3);
    if (cardValue == 0) return false;
    var count = 0;
    for (var i = 0; i < paiGowNumArray.length; i++) {
        if (paiGowNumArray[i] != cardValue) count += paiGowNumArray[i];
    }
    return (count == 5) ? true : false;
};

CFSModeRuleManager.isRuleTypeLeftSame = function (paiGowNumArray) {
    var count = 0;
    var cardValue = CFSModeRuleManager.cardValueInArray(paiGowNumArray, 4);
    if (cardValue == 0) return false;
    for (var i = 0; i < paiGowNumArray.length; i++) {
        if (paiGowNumArray[i] != cardValue) count += paiGowNumArray[i];
    }
    return (count == cardValue) ? true : false;
};

CFSModeRuleManager.isRuleTypeDivide = function (paiGowNumArray) {
    var count = 0;
    var cardValue = CFSModeRuleManager.cardValueInArray(paiGowNumArray, 3);
    if (cardValue == 0) return false;
    var criterionValue = 0;
    for (var i = 0; i < paiGowNumArray.length; i++) {
        if (paiGowNumArray[i] == cardValue) continue;
        criterionValue = paiGowNumArray[i];
        break;
    }
    return (CFSModeRuleManager.getPaiGowCountInArray(paiGowNumArray, criterionValue) == 3) ? true : false;
};

CFSModeRuleManager.getRuleString = function (ruleType) {
    switch (ruleType) {
        case RULETYPE_NONE:
            return MultiLanguageUtil.getLocalizatedStringForKey(STRING_PAIGOW_RULETYPE_NONE);
            break;
        case RULETYPE_NORMAL:
            return MultiLanguageUtil.getLocalizatedStringForKey(STRING_PAIGOW_RULETYPE_NORMAL);
            break;
        case RULETYPE_LEFT_FIVE:
            return MultiLanguageUtil.getLocalizatedStringForKey(STRING_PAIGOW_RULETYPE_LEFT_FIVE);
            break;
        case RULETYPE_DIVIDE:
            return MultiLanguageUtil.getLocalizatedStringForKey(STRING_PAIGOW_RULETYPE_DIVIDE);
            break;
        case RULETYPE_DOUBLE:
            return MultiLanguageUtil.getLocalizatedStringForKey(STRING_PAIGOW_RULETYPE_DOUBLE);
            break;
        case RULETYPE_LEFT_SAME:
            return MultiLanguageUtil.getLocalizatedStringForKey(STRING_PAIGOW_RULETYPE_LEFT_SAME);
            break;
        case RULETYPE_FIVE:
            return MultiLanguageUtil.getLocalizatedStringForKey(STRING_PAIGOW_RULETYPE_FIVE);
            break;
        case RULETYPE_TTS:
            return MultiLanguageUtil.getLocalizatedStringForKey(STRING_PAIGOW_RULETYPE_TTS);
            break;
        case RULETYPE_CONTINUE:
            return MultiLanguageUtil.getLocalizatedStringForKey(STRING_PAIGOW_RULETYPE_CONTINUE);
            break;
        default:
            break;
    }
    return "UNKNOW ERROR!";
};


CFSModeRuleManager.getHintPaiGowObjects = function (paiGowArray) {
    var count = paiGowArray.length;
    if (count < 3) return null;
    // check left 3 cards
    var tempArray = new Array(paiGowArray[0], paiGowArray[1], paiGowArray[2]);
    if (this.getPaiGowRuleStyle(tempArray) != RULETYPE_NONE) return tempArray;
    // check right 3 cards
    tempArray = new Array(paiGowArray[count - 1], paiGowArray[count - 2], paiGowArray[count - 3]);
    if (this.getPaiGowRuleStyle(tempArray) != RULETYPE_NONE) return tempArray;
    // check left 2 && right 1 cards
    tempArray = new Array(paiGowArray[0], paiGowArray[1], paiGowArray[count - 1]);
    if (this.getPaiGowRuleStyle(tempArray) != RULETYPE_NONE) return tempArray;
    // check left 1 && right 2 cards
    tempArray = new Array(paiGowArray[0], paiGowArray[count - 1], paiGowArray[count - 2]);
    if (this.getPaiGowRuleStyle(tempArray) != RULETYPE_NONE) return tempArray;
    return null;
};

File: ClearFiveStages-iOS/ClearFiveStages/Manager/CFSModeRuleManager.m
#import "CFSModeRuleManager.h"
#import "PaiGowObject.h"
#import "GameDataManager.h"
#import "MultiLanguageUtil.h"

@implementation CFSModeRuleManager

+(RuleType)getPaiGowRuleStyle:(NSMutableArray *) paiGowArray {
	if (paiGowArray.count != 3) return RULETYPE_NONE;
	NSMutableArray *paiGowNumArray = [NSMutableArray array];
	for (int i = 0; i < paiGowArray.count; i++) {
		PaiGowObject *p = [paiGowArray objectAtIndex:i];
		[paiGowNumArray addObject:[NSNumber	numberWithInt:p.pg_preNum]];
		[paiGowNumArray addObject:[NSNumber	numberWithInt:p.pg_postNum]];
	}
	if ([self isRuleTypeContinue:paiGowNumArray]) return RULETYPE_CONTINUE;
	if ([self isRuleTypeTTS:paiGowNumArray]) return RULETYPE_TTS;
	if ([self isRuleTypeFive:paiGowNumArray]) return RULETYPE_FIVE;
	if ([self isRuleTypeDouble:paiGowNumArray]) return RULETYPE_DOUBLE;
	if ([self isRuleTypeDivide:paiGowNumArray]) return RULETYPE_DIVIDE;
	if ([self isRuleTypeLeftSame:paiGowNumArray]) return RULETYPE_LEFT_SAME;	
	if ([self isRuleTypeLeftFive:paiGowNumArray]) return RULETYPE_LEFT_FIVE;
	if ([self isRuleTypeNormal:paiGowNumArray]) return RULETYPE_NORMAL;
	return RULETYPE_NONE;
}

+(int) getPaiGowCountInArray:(NSMutableArray *)paiGowArray withValue:(int)cardValue {
	int count = 0;
	for (int i = 0; i < paiGowArray.count; i++) {
		int tmpVal = [[paiGowArray objectAtIndex:i] intValue];
		if (tmpVal == cardValue) count++;
	}
	return count;
}
 
+(int) cardValueInArray:(NSMutableArray *)paiGowArray withCount:(int)cardCount {
	for (int i = 0; i < paiGowArray.count; i++) {
		int cardValue = [paiGowArray[i] intValue];
		int count = [self getPaiGowCountInArray:paiGowArray withValue:cardValue];
		if (count == cardCount) return cardValue;
	}
	return 0;																					
}

+(bool)isRuleTypeTTS:(NSMutableArray *) paiGowNumArray {
	if ([self getPaiGowCountInArray:paiGowNumArray withValue:2] != 2) return NO;
	if ([self getPaiGowCountInArray:paiGowNumArray withValue:3] != 2) return NO;
	if ([self getPaiGowCountInArray:paiGowNumArray withValue:6] != 2) return NO;
	return YES;
}

+(bool)isRuleTypeDouble:(NSMutableArray *) paiGowNumArray {
	if ([self getPaiGowCountInArray:paiGowNumArray withValue:1] != 2) {
		if ([self getPaiGowCountInArray:paiGowNumArray withValue:4] != 2) return NO;
		if ([self getPaiGowCountInArray:paiGowNumArray withValue:5] != 2) return NO;
		if ([self getPaiGowCountInArray:paiGowNumArray withValue:6] != 2) return NO;
		return YES;
	}
	if ([self getPaiGowCountInArray:paiGowNumArray withValue:2] != 2) {
		if ([self getPaiGowCountInArray:paiGowNumArray withValue:4] != 2) return NO;
		if ([self getPaiGowCountInArray:paiGowNumArray withValue:5] != 2) return NO;
		if ([self getPaiGowCountInArray:paiGowNumArray withValue:6] != 2) return NO;
		return YES;
	}
	if ([self getPaiGowCountInArray:paiGowNumArray withValue:3] != 2) {	
		if ([self getPaiGowCountInArray:paiGowNumArray withValue:4] != 2) return NO;
		if ([self getPaiGowCountInArray:paiGowNumArray withValue:5] != 2) return NO;
		if ([self getPaiGowCountInArray:paiGowNumArray withValue:6] != 2) return NO;
		return YES;
	}
	return YES;
}

+(bool)isRuleTypeFive:(NSMutableArray *) paiGowNumArray {
	for (int i = 0; i < paiGowNumArray.count; i++) {	
		int tmpVal = [paiGowNumArray[i] intValue];
		int count = [self getPaiGowCountInArray:paiGowNumArray withValue:tmpVal];	
		if (count == 5) return YES;
	}
	return NO;
}

+(bool)isRuleTypeContinue:(NSMutableArray *) paiGowNumArray {
	for (int i = 1; i < 7; i++) {	
		int index = [paiGowNumArray indexOfObject:[NSNumber numberWithInt:i]];
		if (index == NSNotFound) return NO;
	}
	return YES;
}

+(bool)isRuleTypeNormal:(NSMutableArray *) paiGowNumArray {
	int cardValue = [self cardValueInArray:paiGowNumArray withCount:3];
	if (cardValue == 0) return NO;
	int count = 0;
	for (int i = 0; i < paiGowNumArray.count; i++) {
		int val = [paiGowNumArray[i] intValue];
		if (val != cardValue) count += val;
	}
	int criterionValue = ([GameDataManager sharedManager].gameLevel == GAMELEVEL_EASY) ? 13 : 14;
	return (count >= criterionValue) ? YES : NO;
}

+(bool)isRuleTypeLeftFive:(NSMutableArray *) paiGowNumArray {
	int cardValue = [self cardValueInArray:paiGowNumArray withCount:3];
	if (cardValue == 0) return NO;
	int count = 0;
	for (int i = 0; i < paiGowNumArray.count; i++) {
		int val = [paiGowNumArray[i] intValue];
		if (val != cardValue) count += val;
	}
	return (count == 5) ? YES : NO;
}

+(bool)isRuleTypeLeftSame:(NSMutableArray *) paiGowNumArray {
	int count = 0;
	int cardValue = [self cardValueInArray:paiGowNumArray withCount:4];
	if (cardValue == 0) return NO;
	for (int i = 0; i < paiGowNumArray.count; i++) {
		int val = [paiGowNumArray[i] intValue];
		if (val != cardValue) count += val;
	}
	return (count == cardValue) ? YES : NO;
}

+(bool)isRuleTypeDivide:(NSMutableArray *) paiGowNumArray {
	int cardValue = [self cardValueInArray:paiGowNumArray withCount:3];
	if (cardValue == 0) return NO;
    NSMutableArray *paiGowNumTempArray = [NSMutableArray arrayWithArray:paiGowNumArray];
	[paiGowNumTempArray removeObject:[NSNumber numberWithInt:cardValue]];
	return ([self getPaiGowCountInArray:paiGowNumTempArray withValue:[paiGowNumTempArray[0] intValue]] == 3) ? YES : NO;
}

+(NSMutableArray *) getHintPaiGowObjects:(NSMutableArray *) paiGowArray {
	int count = paiGowArray.count;
	if (count < 3) return nil;
	// check left 3 cards
	NSMutableArray *tempArray = [NSMutableArray arrayWithObjects:paiGowArray[0], paiGowArray[1], paiGowArray[2], nil];
	if ([self getPaiGowRuleStyle:tempArray] != RULETYPE_NONE) return tempArray;
	if (count == 3) return nil;
	// check right 3 cards
	tempArray = [NSMutableArray arrayWithObjects:paiGowArray[-1], paiGowArray[-2], paiGowArray[-3], nil];
	if ([self getPaiGowRuleStyle:tempArray] != RULETYPE_NONE) return tempArray;	
	// check left 2 && right 1 cards
	tempArray = [NSMutableArray arrayWithObjects:paiGowArray[0], paiGowArray[1], paiGowArray[-1], nil];
	if ([self getPaiGowRuleStyle:tempArray] != RULETYPE_NONE) return tempArray;		
	// check left 1 && right 2 cards
	tempArray = [NSMutableArray arrayWithObjects:paiGowArray[0], paiGowArray[-1], paiGowArray[-2], nil];
	if ([self getPaiGowRuleStyle:tempArray] != RULETYPE_NONE) return tempArray;
	return nil;
}

+(NSString *) getRuleString:(RuleType) ruleType { 
	switch (ruleType) {
		case RULETYPE_NONE:
			return [MultiLanguageUtil getLocalizatedStringForKey:STRING_PAIGOW_RULETYPE_NONE];
			break;
		case RULETYPE_NORMAL:
			return [MultiLanguageUtil getLocalizatedStringForKey:STRING_PAIGOW_RULETYPE_NORMAL];		
			break;
		case RULETYPE_LEFT_FIVE:
			return [MultiLanguageUtil getLocalizatedStringForKey:STRING_PAIGOW_RULETYPE_LEFT_FIVE];		
			break;
		case RULETYPE_DIVIDE:
			return [MultiLanguageUtil getLocalizatedStringForKey:STRING_PAIGOW_RULETYPE_DIVIDE];
			break;
		case RULETYPE_DOUBLE:
			return [MultiLanguageUtil getLocalizatedStringForKey:STRING_PAIGOW_RULETYPE_DOUBLE];
			break;
		case RULETYPE_LEFT_SAME:
			return [MultiLanguageUtil getLocalizatedStringForKey:STRING_PAIGOW_RULETYPE_LEFT_SAME];
			break;
		case RULETYPE_FIVE:
			return [MultiLanguageUtil getLocalizatedStringForKey:STRING_PAIGOW_RULETYPE_FIVE];
			break;
		case RULETYPE_TTS:
			return [MultiLanguageUtil getLocalizatedStringForKey:STRING_PAIGOW_RULETYPE_TTS];
			break;
		case RULETYPE_CONTINUE:
			return [MultiLanguageUtil getLocalizatedStringForKey:STRING_PAIGOW_RULETYPE_CONTINUE];
			break;
		default:
			break;
	}
	return [NSString stringWithFormat:@""];
}
@end

#if __has_feature(objc_subscripting) 
#import <objc/runtime.h>
@implementation NSArray (NegativeIndexes)
+ (void)load {
    method_exchangeImplementations(
        class_getInstanceMethod(self, @selector(objectAtIndexedSubscript:)),
        class_getInstanceMethod(self, @selector(BLC_objectAtIndexedSubscript:))
    );
}
- (id)BLC_objectAtIndexedSubscript:(NSInteger)idx {
    if (idx < 0) idx = [self count] + idx;
    return [self BLC_objectAtIndexedSubscript:idx];
}
@end
 
@implementation NSMutableArray (NegativeIndexes)
+ (void)load {
    method_exchangeImplementations(
        class_getInstanceMethod(self, @selector(setObject:atIndexedSubscript:)),
        class_getInstanceMethod(self, @selector(BLC_setObject:atIndexedSubscript:))
    );
}
- (void)BLC_setObject:(id)anObject atIndexedSubscript:(NSInteger)idx {
    if (idx < 0) idx = [self count] + idx;
    [self BLC_setObject:anObject atIndexedSubscript:idx];
}
@end
#endif
